content/Development/qtest.md

```diff
@@ -0,0 +1,224 @@
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing,
+  software distributed under the License is distributed on an
+  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License. -->
+
+# QFile Query Unit Test
+
+QFile Query Unit Test is a JUnit-based integration test suite for Apache Hive. Developers write any SQL; the testing framework runs it and verifies the result and output.
+
+{{< toc >}}
+
+## Tutorial: How to add a new test case
+
+### Preparation
+
+You have to compile Hive's source codes ahead of time.
+
+```sh
+$ cd /path/to/hive
+$ mvn clean install -DskipTests
+$ cd itests
+$ mvn clean install -DskipTests
+```
+
+### Add a QFile
+
+A QFile includes a set of SQL statements that you want to test. Typically, we should put a new file in `ql/src/test/queries/clientpositive`.
+
+Let's say you created the following file.
+
+```sh
+$ cat ql/src/test/queries/clientpositive/aaa.q
+SELECT 1;
+```
+
+### Generate a result file
+
+You can generate the expected output using JUnit.
+
+```sh
+$ cd /path/to/hive/itests/qtest
+$ mvn test -Dtest=TestMiniLlapLocalCliDriver -Dtest.output.overwrite=true -Dqfile=aaa.q
+...
+$ cat ql/src/test/results/clientpositive/llap/aaa.q.out
+PREHOOK: query: SELECT 1
+PREHOOK: type: QUERY
+PREHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+POSTHOOK: query: SELECT 1
+POSTHOOK: type: QUERY
+POSTHOOK: Input: _dummy_database@_dummy_table
+#### A masked pattern was here ####
+1
+```
+
+### Verify the result file
+
+You can ensure the generated result file is correct by rerunning the test case without `test.output.overwrite=true`.
+
+```sh
+$ mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile=aaa.q
+```
+
+## Helpful magic comments
+
+### Using test data
+
+Adding `--! qt:dataset:{table name}`, Query Unit Test automatically sets up a test table. [You can find the table definitions here](https://github.com/apache/hive/tree/master/data/files/datasets).
+
+```sql
+--! qt:dataset:src
+SELECT * FROM src;
+```
+
+### Mask non-deterministic outputs
+
+Some test cases might generate non-deterministic results. You can mask the non-deterministic part using a special comment prefixed with `--! qt:replace:`.
+
+For example, the result of `CURRENT_DATE` changes every day. Using the comment, the output will be `non-deterministic-output #Masked#` , which is stable.
+
+```sql
+--! qt:replace:/(non-deterministic-output\s)[0-9]{4}-[0-9]{2}-[0-9]{2}/$1#Masked#/
+SELECT 'non-deterministic-output', CURRENT_DATE();
+```
+
+## Commandline options
+
+### Run multiple test cases
```

Comment:
The most basic thing which is to run a single query file seems to be missing from this page. I think we should begin this doc from there. It's kinda present in `Verify the result file` section but it should be more explicit.